### PR TITLE
[Perf] Move argmax from host to device for prefill for Falcon7b=B; +10–15% throughput

### PR DESCRIPTION
**Pull request by aliaksei-sala**
### Ticket
[[Perf] Move argmax from host to device for prefill for Falcon7b=B; +10–15% throughput](https://github.com/tenstorrent/tt-metal/issues/30883)

### Problem description
During prefill, we were transferring full logits back to the host solely to compute argmax. 
That introduced:
- Extra PCIe bandwidth and latency
- CPU work on a latency-sensitive path
- Synchronization stalls between device and host

Computing argmax on-device lets us return only token IDs (or a tiny top-k), shrinking transfers and removing a CPU bottleneck.

### What's changed
Move the argmax computation from the host CPU to the device during prefill stage for Falcon7B.

### Performance impact

No changes for single-card tests

Suite / Model | Config | Metric | Actual | Expected | Δ %
-- | -- | -- | -- | -- | --
Galaxy Falcon-7B demos | perf_mode_128 | Prefill t/s | 66,485 | 32,900 | +102%
Galaxy Falcon-7B demos | perf_mode_1024 | Prefill t/s | 104,109 | 32,000 | +225%
Galaxy Falcon-7B demos | perf_mode_1024 | Decode t/s | 11,349 | 11,776 | −3.6%
Galaxy Falcon-7B demos | perf_mode_1024 | Decode t/s/u | 11.08 | 11.5 | −3.6%
Galaxy Falcon-7B demos | perf_mode_2048 | Prefill t/s | 84,600 | 31,100 | +172%
T3000 (T3K) Falcon-7B | perf_mode_128 | Prefill t/s | 16,315 | 10,900 | +49.7%

### Checklist
- [x] [single-card-demo-tests.yaml](https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/single-card-demo-tests.yaml) 
-- [Result](https://github.com/tenstorrent/tt-metal/actions/runs/18956822978) 128 seq len
-- [R2](https://github.com/tenstorrent/tt-metal/actions/runs/18974346050)  1024 seq len
-- [R3](https://github.com/tenstorrent/tt-metal/actions/runs/18977019209) 2048 seq len
-- [R4](https://github.com/tenstorrent/tt-metal/actions/runs/18981174093) all prefill values are up-to-date
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes. 
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) [Results](https://github.com/tenstorrent/tt-metal/actions/runs/18666474128) [R2](https://github.com/tenstorrent/tt-metal/actions/runs/18700308304)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) [Results](https://github.com/tenstorrent/tt-metal/actions/runs/18666496183) [R2](https://github.com/tenstorrent/tt-metal/actions/runs/18700321227) [R3](https://github.com/tenstorrent/tt-metal/actions/runs/18789399079)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [x] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable) [Result](https://github.com/tenstorrent/tt-metal/actions/runs/18724844033)
- [x] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work 
-- [Result](https://github.com/tenstorrent/tt-metal/actions/runs/18724865494) 
-- [R2](https://github.com/tenstorrent/tt-metal/actions/runs/18852150892)
- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main) 
-- [Result](https://github.com/tenstorrent/tt-metal/actions/runs/18724886385)
- [x] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release) 
-- [Result](https://github.com/tenstorrent/tt-metal/actions/runs/18724902778) 
-- [R2](https://github.com/tenstorrent/tt-metal/actions/runs/18789454218) 
-- [R3](https://github.com/tenstorrent/tt-metal/actions/runs/18852137789)
-- [R4](https://github.com/tenstorrent/tt-metal/actions/runs/18862441771)
-- [R5](https://github.com/tenstorrent/tt-metal/actions/runs/18917622088) 2048 seq len
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by skhorasganiTT — COMMENTED** [review]
Thanks! Could you please run all the demo pipelines (single card, t3k, galaxy) and to check if perf targets need to be updated?

**Comment by aliaksei-sala** [comment]
Hi, @skhorasganiTT 
Performance targets have been updated. 
I’ll run a few tests first to verify reproducibility before moving forward

File: models/demos/tg/falcon7b/demo_tg.py [hunk 1]
```diff
@@ -76,9 +76,9 @@ def test_demo_multichip(
                 2048: {"prefill_t/s": 20400, "decode_t/s/u": 6.60},
             },
             "6U": {
-                128: {"prefill_t/s": 32900, "decode_t/s/u": 11.60},
-                1024: {"prefill_t/s": 32000, "decode_t/s/u": 11.50},
-                2048: {"prefill_t/s": 31100, "decode_t/s/u": 10.97},
+                128: {"prefill_t/s": 66485, "decode_t/s/u": 11.60},
+                1024: {"prefill_t/s": 104109, "decode_t/s/u": 11.50},
+                2048: {"prefill_t/s": 84600, "decode_t/s/u": 10.97},
             },
```
Comment (ayerofieiev-tt) [inline models/demos/tg/falcon7b/demo_tg.py]: this update is based on CI numbers, right?

Comment (aliaksei-sala) [reply models/demos/tg/falcon7b/demo_tg.py]: Yes, I’ve added links to the CI runs in the PR checklist, where you can find these numbers.

T3K CI : https://github.com/tenstorrent/tt-metal/actions/runs/18724902778/job/53409844279
Galaxy CI : https://github.com/tenstorrent/tt-metal/actions/runs/18724865494/job/53409311467

Comment (skhorasganiTT) [reply models/demos/tg/falcon7b/demo_tg.py]: @aliaksei-sala please set the new prefill targets to be 5% lower than the values observed on CI since there is some variability. Also fyi, there is an issue regarding decode perf variability for T3K-Falcon7B-CI: https://github.com/tenstorrent/tt-metal/issues/30679, if you see decode perf failures please ignore as your PR only affects prefill.

**Comment by skhorasganiTT** [comment]
@aliaksei-sala I would also recommend temporarily commenting out non-falcon7b tests in the CI workflow scripts to speed up testing and save compute resources (for single card - https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/single-card-demo-tests-impl.yaml#L90, for t3k - https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/t3000-demo-tests-impl.yaml#L34, for glx - https://github.com/tenstorrent/tt-metal/blob/main/.github/workflows/galaxy-demo-tests-impl.yaml#L35).

**Comment by aliaksei-sala** [comment]
@skhorasganiTT, 

- Single-card device perf has never failed which is **odd** since my local N150 runs showed a performance improvement(results are attached to github issue).
- Galaxy tests with new expected prefill numbers are **passed**. 
- T3K pipeline failed before I reached 2048 seq len with : 
`FAILED models/demos/t3000/falcon7b/demo_t3000.py::test_demo_multichip[wormhole_b0-user_input0-8-perf_mode_128_stochastic_verify] 
Read unexpected run_mailbox value: 0x40 (expected 0x80 or 0x0)`

I’ve commented out max_seq_len = 128/1024 to get prefill numbers for a 2048 sequence length. I’ll update today.

**Review by skhorasganiTT — APPROVED** [review]
@aliaksei-sala could you please also link the single card demo (https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI run in the PR description? 
I noticed in your previous comment you said single card device perf instead of demo, just wanted to make sure you ran the right pipeline and that the targets there are okay.

**Review by skhorasganiTT — APPROVED** [review]
